synapse/push/clientformat.py

```diff
@@ -18,28 +18,41 @@
 from synapse.push.rulekinds import PRIORITY_CLASS_INVERSE_MAP, PRIORITY_CLASS_MAP
 from synapse.types import UserID
 
+from .baserules import FilteredPushRules, PushRule
+
 
 def format_push_rules_for_user(
-    user: UserID, ruleslist: List
+    user: UserID, ruleslist: FilteredPushRules
 ) -> Dict[str, Dict[str, list]]:
     """Converts a list of rawrules and a enabled map into nested dictionaries
     to match the Matrix client-server format for push rules"""
 
-    # We're going to be mutating this a lot, so do a deep copy
-    ruleslist = copy.deepcopy(ruleslist)
-
     rules: Dict[str, Dict[str, List[Dict[str, Any]]]] = {
         "global": {},
         "device": {},
     }
 
     rules["global"] = _add_empty_priority_class_arrays(rules["global"])
 
-    for r in ruleslist:
-        template_name = _priority_class_to_template_name(r["priority_class"])
+    for r, enabled in ruleslist:
+        template_name = _priority_class_to_template_name(r.priority_class)
+
+        rulearray = rules["global"][template_name]
+
+        template_rule = _rule_to_template(r)
+        if not template_rule:
+            continue
```

Comment:
This change looks unrelated to the changes to attrs. Is this avoiding doing work that we don't care about if the template rule isn't found anyway?

Reply:
I found it easier to get the `template_rule` early and then update that, rather than what was here before which deep copied the push rules and mutated them in place.

Reply:
That makes sense -- I guess my thought is this isn't inherent to the change to structures. 👍

Reply:
I think this caused #14348 because the cleanup logic ("# Remove internal stuff" below)is not longer applied before `_rule_to_template` is called.